### DISCUSSION TITLE
Is it possible to use Kubebuilder with custom Dockerfile in root directory?

### DISCUSSION THREAD

**Discussion by nyrareddy** [kubernetes-sigs/kubebuilder#2703] [Support] [answered]
I need to add a new CRD as per new requirements but when I try:
```
kubebuilder edit --multigroup
kubebuilder create api --group groupname --version version --kind kindName
```
Then I get this error:
```
failed to edit project: unable to scaffold with "go.kubebuilder.io/v2": can't find "COPY api/ api/"
```
I have a need for using a custom Dockerfile in the project root but want to continue using Kubebuilder. Also worth mentioning might be that the `main.go` files are in a new `cmd` directory. Can you please suggest a way to use kubebuilder in this project?

**Comment by FillZpp** [comment]
I think the problem is `go.kubebuilder.io/v2` doesn't support `multigroup` so it could only recognize `api/` directory. You may have to update to `go.kubebuilder.io/v3`. See [APIs migration from v2 to v3](https://book.kubebuilder.io/migration/migration_guide_v2tov3.html#migrate-the-apis)

**Reply by camilamacedo86** [reply]
@FillZpp multigroup is supported since 2.x and in this way with go/v2. The problem here is the customizations/changes done in the Dockerfile, see: ` "failed to edit project: unable to scaffold with "go.kubebuilder.io/v2": can't find "COPY api/ api/""` the plugin did not find it in the file to do the replace.

**Comment by camilamacedo86** [comment, marked as answer]
@nyrareddy, 

Multi-group is supported by go/v2. The problem here is that you update the Dockerfile and when we are trying to do the changes on the scaffolds we will look for "COPY api/ api/" and it is not found. 

Therefore, yes you can change the scaffold as please you. I'd not recommend you to do it unless you need it and you know what you are doing because if you deviate too much from the scaffold proposed then some actions can no longer work and you will make the process to upgrade your project very complex. Therefore, you might be impacted and you will no longer be able to take advantage of helpers/plugins and etc that can be provided.  

See: 

<img width="865" alt="Screenshot 2022-07-27 at 11 39 06" src="https://user-images.githubusercontent.com/7708031/181227617-88adcfd3-aad7-4236-a5d3-03cfc0bd3f20.png">

From: https://book.kubebuilder.io/migration/v2vsv3.html#project-customizations

IMPORTANT: I would like to encourage you to move forward and migrate your project since you are using go/v2. See:

> 📢 Deprecations
> Legacy layout "Kubebuilder 2.x" provided via go/v2 plugin is deprecated
> Since 28 Apr 2021, the default layout produced by Kubebuilder changed and is done via the go/v3.
> 
> Who is affected?
> Only affects who has been scaffolding the projects with --plugins=go/v2 or built the project with Kubebuider 2.x and did not migrate their project for Kubebuilder 3.x (go/v3).
> 
> Why this deprecation is required?
> The go/v2 plugin cannot scaffold projects which use CRDs and/or Webhooks v1 API version. The go/v2 plugin produces scaffolds with the v1beta1 API version for CRDs and Webhooks which was deprecated in Kubernetes 1.16 and removed in 1.22.
> 
> Note that this plugin was kept to ensure backwards compatibility with projects that were scaffolded with the old "Kubebuilder 2.x" layout and does not properly work with the new plugin ecosystem that was introduced with Kubebuilder 3.0.0 [More info](https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/plugins/plugins.md).
> 
> What is recommended?
> We encourage you to migrate your project. For further information check the [Migration guide](https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/migration/manually_migration_guide_v2_v3.md).
> 
> 💡 You can check samples using this plugin by looking at the project-v2 directory under the [testdata](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata) on the root directory of the Kubebuilder project and compare with the project-v3 to check the changes.
> 
> Notes:
> 
> ⚠️ The flags --crd-version and --webhook-version provided via go/v3 plugin (default scaffold) are deprecated since kubebuilder release [v3.3.0](https://github.com/kubernetes-sigs/kubebuilder/releases/tag/v3.3.0). https://github.com/kubernetes-sigs/kubebuilder/pull/2340.
> 
> We do not encourage you to use these options since projects produced with CRD/Webhooks v1beta1 version cannot work on Kubernetes clusters >= 1.22 and by using the API v1 (default scaffold when the flags are not used) you can produce projects which can work on K8s clusters >= 1.16.

From: https://github.com/kubernetes-sigs/kubebuilder/releases/tag/v3.5.0

**Comment by camilamacedo86** [comment]
Closing since it is answered.
